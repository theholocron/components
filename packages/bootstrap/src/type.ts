import * as React from "react";

export interface WithChildren {
	children?: React.ReactNode;
}
