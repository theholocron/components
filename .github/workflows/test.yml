name: Test

on:
  push

jobs:
  types:
    name: Check build and types
    runs-on: ubuntu-latest
    strategy:
      matrix:
        package:
          - packages/bootstrap
    steps:
      - uses: actions/checkout@v4
        name: Checkout repository
        with:
          fetch-depth: 0

      - uses: actions/setup-node@v4
        name: Setup Node.js @v20
        with:
          cache: npm
          node-version: 20.x

      - run: npm ci
        name: Install dependencies
        env:
          HUSKY: 0

      - run: npm rebuild typescript && npm run build:lib && npm run types -- --project tsconfig.build.json
        name: Build the library and types
        working-directory: ${{ matrix.package }}

  coverage:
    name: Run tests and collect coverage
    runs-on: ubuntu-latest
    strategy:
      matrix:
        package:
          - packages/bootstrap
    steps:
      - uses: actions/checkout@v4
        name: Checkout repository
        with:
          fetch-depth: 0

      - uses: actions/setup-node@v4
        name: Setup Node.js @v20
        with:
          cache: npm
          node-version: 20.x

      - run: npm ci
        name: Install dependencies
        env:
          HUSKY: 0

      - run: npx playwright install --with-deps
        name: Install Playwright

      - run: npm run build:storybook -- --quiet --output-dir=docs
        name: Build Storybook
        working-directory: ${{ matrix.package }}

      - run: |
          npx concurrently -k -s first -n "SB,TEST" -c "magenta,blue" \
          "npx http-server docs --port 6006 --silent" \
          "npx wait-on tcp:6006 && npm run test:storybook -- --coverage --junit --ci"
        name: Run tests with coverage
        working-directory: ${{ matrix.package }}

      - uses: codecov/codecov-action@v4
        name: Upload results to Codecov
        with:
          files: ./${{ matrix.package }}/coverage/coverage-final.json
          flags: ${{ matrix.package }}
          token: ${{ secrets.CODECOV_TOKEN }}

      - uses: codecov/test-results-action@v1
        name: Upload test results to Codecov
        if: ${{ !cancelled() }}
        with:
          working-directory: ${{ matrix.package }}
          token: ${{ secrets.CODECOV_TOKEN }}

  # Run visual and composition tests with Chromatic
  visual-and-composition:
    name: Test Visual and Composition
    runs-on: ubuntu-latest
    strategy:
      matrix:
        package:
          - packages/bootstrap
    steps:
      - uses: actions/checkout@v4
        name: Checkout repository
        with:
          fetch-depth: 0

      - uses: actions/setup-node@v4
        name: Setup Node.js @v20
        with:
          cache: npm
          node-version: 20.x

      - run: npm ci
        name: Install dependencies
        env:
          HUSKY: 0

      - uses: chromaui/action@v1
        name: Publish to Chromatic
        with:
          buildScriptName: build:storybook
          projectToken: ${{ secrets.CHROMATIC_PROJECT_TOKEN }}
          token: ${{ secrets.GITHUB_TOKEN }}
          workingDir: ${{ matrix.package }}
        env:
          NEXT_PUBLIC_ENVIRONMENT: "development"
          STORYBOOK_NEXT_PUBLIC_ENVIRONMENT: "development"

  # Run interaction and accessibility tests
  interaction-and-accessibility:
    name: Test Interactions and Accessibility
    runs-on: ubuntu-latest
    strategy:
      matrix:
        package:
          - packages/bootstrap
    steps:
      - uses: actions/checkout@v4
        name: Checkout repository
        with:
          fetch-depth: 0

      - uses: actions/setup-node@v4
        name: Setup Node.js @v20
        with:
          cache: npm
          node-version: 20.x

      - run: npm ci
        name: Install dependencies
        env:
          HUSKY: 0

      - run: npx playwright install --with-deps
        name: Install Playwright

      - run: npm run build:storybook -- --quiet --output-dir=docs
        name: Build Storybook
        working-directory: ${{ matrix.package }}

      - run: |
          npx concurrently -k -s first -n "SB,TEST" -c "magenta,blue" \
          "npx http-server docs --port 6006 --silent" \
          "npx wait-on tcp:6006 && npm run test:storybook -- --ci"
        name: Serve Storybook and run tests
        working-directory: ${{ matrix.package }}

  # Run user flow tests with Cypress
  user-flow:
    name: Test User Flow
    runs-on: ubuntu-latest
    strategy:
      fail-fast: false # https://github.com/cypress-io/github-action/issues/48
      matrix:
        containers: [1, 2] # Uses 2 parallel instances
        package:
          - packages/bootstrap
    steps:
      - uses: actions/checkout@v4
        name: Checkout repository
        with:
          fetch-depth: 0

      - uses: actions/setup-node@v4
        name: Setup Node.js @v20
        with:
          cache: npm
          node-version: 20.x

      - run: npm ci
        name: Install dependencies
        env:
          HUSKY: 0

        # Uses the official Cypress GitHub action https://github.com/cypress-io/github-action
      - uses: cypress-io/github-action@v6
        name: Cypress run
        with:
          # Starts web server for E2E tests - replace with your own server invocation
          # https://docs.cypress.io/guides/continuous-integration/introduction#Boot-your-server
          start: npm run dev
          wait-on: "http://localhost:3000" # Waits for above
          # Records to Cypress Cloud
          # https://docs.cypress.io/guides/cloud/projects#Set-up-a-project-to-record
          record: true
          parallel: true # Runs test in parallel using settings above
          working-directory: ${{ matrix.package }}
        env:
          # For recording and parallelization to work you must set your CYPRESS_RECORD_KEY
          # in GitHub repo → Settings → Secrets → Actions
          CYPRESS_RECORD_KEY: ${{ secrets.CYPRESS_RECORD_KEY }}
          # Creating a token https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          HUSKY: 0
