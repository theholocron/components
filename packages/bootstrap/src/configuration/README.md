# Configuration Component

A React Provider component for storing an applications configuration.

## Usage

```typescript
import { Configuration } from "@theholocron/bootstrap";
import * as React from "react";

interface AppProps {
    Component: React.ReactNode;
    pageProps?: any;
}

export default function App (props: AppProps) {
    const { Component, pageProps } = props;
    const conf = {
        description: "My app description";
        id: "app-name";
        name: "App name";
        slug: "app-name-ui";
        title: "App Name UI";
        version: 2023.1.0;
    };

    return (
        <Configuration.Provider conf={conf}>
            <Component {...pageProps} />
        </Configuration.Provider>
    );
}
```

### Hook

There is a hook that comes along with the library for retrieving the applications configuration.

```typescript
import { useConfiguration } from "@theholocron/bootstrap";

function Component () {
    const conf = useConfiguration();

    return (
        <div>
            <h3>Configuration</h3>
            <pre>{JSON.stringify(conf, null, 4)}</pre>
        </div>
    );
}
```

## Complete Configuration

```json
{
    application: {
        description: "This is a description for a mock application.",
        id: "mock-app-id",
        name: "A Mock Application",
        get slug (): string {
            return this.id;
        },
        title: "The title of the application, which maps to <title />",
        version: <appVersion || pkg.version>,
    },
    fetch: {
        conf: {
            ...otherAxiosConfig // https://github.com/axios/axios/blob/v1.x/index.d.ts#L314,
        },
        swr: {
            ...otherSWRConfig // https://swr.vercel.app/docs/api#options
        },
    },
}
```
