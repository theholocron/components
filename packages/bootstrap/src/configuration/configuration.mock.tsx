import { useConfiguration } from "./index.ts";

export const mockConf = {
	application: {
		description: "This is a description for a mock application.",
		id: "mock-app-id",
		name: "A Mock Application",
		get slug(): string {
			return `${this.id}-slug`;
		},
	},
	// fetch: {},
};

export function TestComponent() {
	const conf = useConfiguration();

	return (
		<div data-testid="mock-configuration">
			<pre>{JSON.stringify(conf, null, 4)}</pre>
		</div>
	);
}
