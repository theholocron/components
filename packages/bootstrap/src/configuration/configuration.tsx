import * as str from "@theholocron/utils-string";
import * as React from "react";
import type { WithChildren } from "../type.ts";

export interface ApplicationConf {
	description?: string;
	id: string;
	name: string;
	slug?: string;
	title?: string;
	version?: string;
}

export interface IConfiguration {
	application: ApplicationConf;
	// fetch?: FetchConf;
}

const Context = React.createContext<IConfiguration | undefined>(undefined);
Context.displayName = "@theholocron/bootstrap/Configuration";

export interface ConfigurationProps extends WithChildren {
	conf: IConfiguration;
}

export function Provider(props: ConfigurationProps) {
	const conf = {
		...props.conf,
		application: {
			...props.conf.application,
			slug: str.toKebabCase(
				props.conf.application?.slug || props.conf.application.id
			),
			title:
				props.conf.application?.title ??
				str.toTitleCase(props.conf.application?.name),
		},
	};

	const value = React.useMemo(() => conf, [conf]);

	return <Context.Provider value={value}>{props.children}</Context.Provider>;
}
Provider.displayName = "@theholocron/bootstrap/Configuration";

export function useConfiguration(): IConfiguration {
	const context = React.useContext(Context);

	if (context === undefined) {
		throw new Error(
			`useConfiguration must be used within Configuration.Provider!`
		);
	}

	return context;
}
