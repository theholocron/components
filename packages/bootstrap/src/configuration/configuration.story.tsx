import type { Meta, StoryObj } from "@storybook/react";
import { expect, within } from "@storybook/test";
import * as MockConfiguration from "./configuration.mock";
import Configuration, { useConfiguration } from "./index";

const meta = {
	argTypes: {
		conf: {
			control: "object",
			defaultValue: MockConfiguration.mockConf,
			description: "The applications configuration.",
		},
	},
	component: Configuration.Provider,
	title: "Configuration",
} satisfies Meta<typeof Configuration.Provider>;
export default meta;
type Story = StoryObj<typeof meta>;

export const Default = {
	render: (args) => (
		<Configuration.Provider {...args}>
			<MockConfiguration.TestComponent />
		</Configuration.Provider>
	),
	args: {
		conf: MockConfiguration.mockConf,
	},
	play: async ({ canvasElement }) => {
		const canvas = within(canvasElement);

		// Check if the initial storage state is as expected
		const initialStorage = await canvas.findByTestId("mock-configuration");
		expect(initialStorage).toBeTruthy();

		const appId = await canvas.findByText(/"id": "mock-app-id"/);
		expect(appId).toBeInTheDocument();

		const appName = await canvas.findByText(/"name": "A Mock Application"/);
		expect(appName).toBeInTheDocument();
	},
} satisfies Story;

// Attempt to use the hook without provider
export const Error = {
	render: () => {
		const Component = () => {
			try {
				useConfiguration();
			} catch (error: unknown) {
				return (
					<div data-testid="error-message">
						{(error as Error)?.message}
					</div>
				);
			}
			return null; // Render nothing if no error
		};

		return <Component />;
	},
	play: async ({ canvasElement }) => {
		const canvas = within(canvasElement);

		const errorMessage = await canvas.findByTestId("error-message");
		expect(errorMessage).toHaveTextContent(
			"useConfiguration must be used within Configuration.Provider!"
		);
	},
} satisfies Story;

// Fallback to ID when slug is not present
export const FallbackSlug = {
	args: {
		conf: {
			...MockConfiguration.mockConf,
			application: {
				...MockConfiguration.mockConf.application,
				name: "",
				id: "fallback-slug",
				slug: "",
			},
		},
	},
	render: Default.render.bind({}),
	play: async ({ canvasElement }) => {
		const canvas = within(canvasElement);
		const appName = await canvas.findByText(/"slug": "fallback-slug"/);
		expect(appName).toBeInTheDocument();
	},
} satisfies Story;

// Fallback when only ID is present
export const FallbackTitle = {
	args: {
		conf: MockConfiguration.mockConf,
	},
	render: Default.render.bind({}),
	play: async ({ canvasElement }) => {
		const canvas = within(canvasElement);
		const appName = await canvas.findByText(
			/"title": "A Mock Application"/
		);
		expect(appName).toBeInTheDocument();
	},
} satisfies Story;
