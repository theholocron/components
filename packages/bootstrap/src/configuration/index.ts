import { Provider } from "./configuration";

export { useConfiguration } from "./configuration";
export type { ApplicationConf, IConfiguration } from "./configuration";

export default { Provider };
