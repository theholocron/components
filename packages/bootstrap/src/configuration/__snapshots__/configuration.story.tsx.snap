// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration Default play-test 1`] = `
<div data-testid="mock-configuration">
  <pre>
    {
    "application": {
        "description": "This is a description for a mock application.",
        "id": "mock-app-id",
        "name": "A Mock Application",
        "slug": "mock-app-id-slug",
        "title": "A Mock Application"
    }
}
  </pre>
</div>
`;

exports[`Configuration Error play-test 1`] = `
<div data-testid="error-message">
  useConfiguration must be used within Configuration.Provider!
</div>
`;

exports[`Configuration FallbackSlug play-test 1`] = `
<div data-testid="mock-configuration">
  <pre>
    {
    "application": {
        "description": "This is a description for a mock application.",
        "id": "fallback-slug",
        "name": "",
        "slug": "fallback-slug",
        "title": ""
    }
}
  </pre>
</div>
`;

exports[`Configuration FallbackTitle play-test 1`] = `
<div data-testid="mock-configuration">
  <pre>
    {
    "application": {
        "description": "This is a description for a mock application.",
        "id": "mock-app-id",
        "name": "A Mock Application",
        "slug": "mock-app-id-slug",
        "title": "A Mock Application"
    }
}
  </pre>
</div>
`;
